Generate per ISA register.rs files in the Rust meta crate

bnjbvr opened the pull request:
Still a WIP: I'm not done with all the ISAs, and there's many small things to change (move things around in other Rust files; change API to make things easier). But this generates an equivalent registers-x86.rs file, at least now.

sunfishcode commented on lib/codegen/meta/src/isa/mod.rs:
```diff
@@ -0,0 +1,399 @@
+            Isa::Arm32 => "arm32",
+            Isa::Arm64 => "arm64",
+        }
+    }
```
I know this is pre-existing code, but could you move this code to a `Display` implementation while here? I just converted https://github.com/CraneStation/cranelift/pull/581.

sunfishcode commented on lib/codegen/meta/src/isa/mod.rs:
```diff
@@ -0,0 +1,399 @@
+    subclasses: Vec<RegClassIndex>,
+    pub toprc: RegClassIndex,
+    pub index: RegClassIndex,
+}
```
Could you sort these fields to more closely follow the order of fields in the `lib/codegen/src/isa/registers.rs` version of `RegClass`, so that it's easier to compare the two?

bjorn3 commented on lib/codegen/meta/src/isa/x86/mod.rs:
```diff
@@ -0,0 +1,40 @@
+pub fn define() -> isa::TargetIsa {
+    let mut isa = isa::TargetIsa::new("x86");
+
+    let int_reg_prefixes = ["rax", "rcx", "rdx", "rbx", "rsp", "rbp", "rsi", "rdi"];
```
If you put `vec!` before the `[`, you dont have to call `.to_vec()` below.

bjorn3 commented on lib/codegen/meta/src/isa/x86/mod.rs:
```diff
@@ -0,0 +1,40 @@
+
+    let builder = isa::RegBankBuilder::new("FlagRegs", "Flag registers", "")
+        .units(1)
+        .reg_names(["rflags"].to_vec())
```
Same here

bnjbvr commented on lib/codegen/meta/src/srcgen.rs:
```diff
@@ -12,24 +12,37 @@ use error;
+        $($middle)*
+        $fmt.indent_pop();
+    }
 }
```
Right, I wasn't even sure of the utility myself. It would have been nice to have a structure IndentedScope, but there's a lifetime issue in there: if we want the `drop()` method to pop by one indent level, then the formatter needs to be mutable in the IndentedScope struct; which means it can't be mutably borrowed between the start of the IndentedScope and the time it's dropped:

```rust
{
  let scope = IndentedScope::new(fmt);
  fmt.write(...); // can't mutate since it's mutably borrowed by scope right above
}
```

So I got back to using this macro instead, which is a small bit nicer than not having it.
If anybody has a better idea for the design of this thing, please chime in!

sunfishcode replied:
Interesting idea -- I think I like it. As a lambda, it should be easier to step through in a debugger than a macro.

bnjbvr commented:
Updated, and doesn't use `u128` anymore by changing a bit how the register class bitmask generation works.

A few notes about the implementation:

1. the documentation of register classes / banks was unused, so I removed it from the structures, but we could keep it, until we figure a better use. (Probably we can just include it as a documentation comment?)
2. It is not obvious from looking at the commits, but the generated files are almost exactly the same as the ones we get with Python. The only difference is the name prefix for the FLAG register banks on all archs, which is "" in my patch while it was "r" with the Python generation; it is not even used in the code, so it doesn't matter that it is different. Other than that, everything is the same, from the indent levels to the trailing commas. I was wondering whether we'd want some more engineering to have the ability (using e.g. environment variables) to choose the generator (Python or Rust) at build time. I don't think it's worth it in the long term, but if it can help approving this PR, I'm happy to do it.

bnjbvr commented on lib/codegen/meta/src/cdsl/regs.rs:
```diff
@@ -0,0 +1,199 @@
+    pub name: &'static str,
+    pub first_unit: u8,
+    pub units: u8,
+    pub names: Option<Vec<&'static str>>,
```
No, not really, good idea to just use an empty vector instead.

bnjbvr commented on lib/codegen/meta/src/gen_registers.rs:
```diff
@@ -0,0 +1,149 @@
+    fn into(self) -> RegUnit {
+        self as RegUnit
+    }
+}"#,
```
Yes, I went back and forth on this. The multi-line meant fewer `format!` macros, but this eye-jumping in the multi-line format makes it less pleasant to read. Reverted back to one per line + indent.

bnjbvr replied:
(this comment was rather an answer to your other comment)

sunfishcode commented:
Looks good!
